### PR TITLE
add batch_size argument to jax.lax.map

### PR DESCRIPTION
**Pull request by cgarciae**
# Changes

* Adds `batch_size` argument to `map`, this allows processing multiple elements per iteration, elements on each batch are  processed in parallel using `vmap`. With these changes `map` can be used as a memory-efficient alternative to `vmap`.
* Fixes `mypy` error.

Original description: Upstreams `batch_vmap` implementation by @shoyer, it has similar semantics to `vmap` but its executed in independent baches using `scan` to reduce memory consumption.

### CODE REVIEW

**Comment by jakevdp** [comment]
Looks like a great start! Cross-referencing a couple related issues: #11319 #18398

File: jax/_src/lax/control_flow/loops.py [hunk 1]
```diff
@@ -2020,6 +2022,97 @@ def map(f, xs):
...
+) -> F:
+  """jax.vmap, but looping when the batch dimension exceeds batch_size.
+
+  Wrapping f with jax.remat() is recommended if you want to use batch_vmap
+  inside gradient calculations. Otherwise, you won't realize any memory savings
+  because JAX will save all the forward activations.
+
+  If your code looks like value_and_grad(mean(vmap(...))) then you probably want
+  to use microbatching instead, which is more efficient because you don't need
+  to use remat. See google3/learning/deepmind/jax/microbatching/
```
Comment (shoyer) [inline jax/_src/lax/control_flow/loops.py:2057]: This should be removed and/or should reference an open source implementation :)

**Comment by Conchylicultor** [comment]
Any update on this ? this came again in cl/606762040

File: jax/_src/lax/control_flow/loops.py [hunk 2]
```diff
@@ -2020,6 +2022,97 @@ def map(f, xs):
...
+def batch_vmap(
+  f: F,
+  in_axes: int | None | Sequence[Any] = 0,
+  out_axes: Any = 0,
+  *,
+  batch_size: int,
+) -> F:
+  """jax.vmap, but looping when the batch dimension exceeds batch_size.
+
+  Wrapping f with jax.remat() is recommended if you want to use batch_vmap
```
Comment (Conchylicultor) [inline jax/_src/lax/control_flow/loops.py:2051]: What's the behavior when the vmap dimension is not divisible by `batch_size` ? Is it an error ? Is the remainder padded (hopefully not) ? ...

It would be nice to document this

**Review by Conchylicultor — CHANGES_REQUESTED** [review]

**Comment by cgarciae** [comment]
Update: cleaned up the implementation a bit, required all `in_axis` dimension to be divisible by `batch_size`, removed the reminders, and improved the docstring.

@shoyer @mattjj  seems like the current implementation doesn't support `None` in the `out_axis`, do we want to support this?

File: jax/_src/lax/control_flow/loops.py (hunk 2 above)
Comment (cgarciae) [reply jax/_src/lax/control_flow/loops.py:2048]: New implementation will yield an error.

**Comment by cgarciae** [comment]
@PhilipVinc `batched_vmap` does sounds appropriate, I'll do the renaming unless someone has a different opinion.

File: jax/_src/lax/control_flow/loops.py (hunk 2 above)
Comment (Conchylicultor) [reply jax/_src/lax/control_flow/loops.py]: 1. In our use-case, there's no guarantee `dim % batch_size == 0`. How difficult would it be to support arbitrary dimension ?
2. Otherwise, the limitation should be documented in the docstring of the function (this doesn't seems to be the case unless I missed something)

File: jax/_src/lax/control_flow/loops.py [hunk 3]
```diff
@@ -2020,6 +2021,107 @@ def map(f, xs):
   _, ys = scan(g, (), xs)
   return ys
 
+
+def _lax_map(f, *xs):
```
[the remark this answers is not in the document]
Comment (cgarciae) [reply jax/_src/lax/control_flow/loops.py:2025]: Maybe its too much for this PR.

[the remark this answers is not in the document]
Comment (froystig) [reply jax/_src/lax/control_flow/loops.py:2025]: @mattjj and I were just wondering the same. Namely, could the API for this be an option on `lax.map`, rather than a top-level function of its own (currently `batched_vmap` in this PR).

File: jax/_src/lax/control_flow/loops.py [hunk 4]
```diff
@@ -2020,6 +2021,107 @@ def map(f, xs):
...
+  because JAX will save all the forward activations.
+
+  If your code looks like value_and_grad(mean(vmap(...))) then you probably want
+  to use microbatching instead, which is more efficient because you don't need
+  to use remat.
+
+  Args:
+    f: A Python callable
+    in_axes: An integer, a pytree of integers, or None. Specifies the axes
+      that should be batched. If None, all axes are batched.
```
Comment (mattjj) [inline jax/_src/lax/control_flow/loops.py:2073]: I think `None` already means something else, i.e. don't batch anything.

Can we just delete this last sentence?

```suggestion
      that should be batched.
```

File: jax/_src/lax/control_flow/loops.py [hunk 5]
```diff
@@ -2020,6 +2021,107 @@ def map(f, xs):
...
+    out_axes: An integer or a pytree of integers. Specifies the axes that
+      should be batched in the output. If None, all axes are batched.
```
Comment (mattjj) [inline jax/_src/lax/control_flow/loops.py:2075]: Same here. I don't think this should be the behavior for `None`, unless I'm misunderstanding.

```suggestion
      should be batched in the output.
```

File: tests/api_test.py [hunk 6]
```diff
@@ -10205,6 +10205,41 @@ def vmap_ref(xs, y):
...
+  def test_batch_map_pytrees(self, batch_size: int):
+    f = lambda x: {'b': x['a'] ** 2}
+    inputs = {'a': np.arange(16)}
+    expected = np.arange(16) ** 2
+
+    outputs = jax.lax.batched_vmap(f, batch_size=batch_size)(inputs)
+    np.testing.assert_array_equal(outputs['b'], expected)
+
+    outputs = jax.lax.batched_vmap(
+      f, batch_size=batch_size, in_axes=({"a": 0},), out_axes={"b": -1}
```
Comment (mattjj) [inline tests/api_test.py:10239]: The in_axes tests here are really limited. We should probably write something combinatorial that checks all cases.

One thing we could do is leave in_axes/out_axes for a follow-up PR, and focus on landing the basic feature first. WDYT?

File: jax/_src/lax/control_flow/loops.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (mattjj) [reply jax/_src/lax/control_flow/loops.py:2025]: More from convo with @froystig: It's kind of analogous to the `unroll` option on `scan`. We're not unrolling here, i.e. we're not stamping out more copies of the same program in sequence, but rather we're adding a batch dim to the body; but still it's computing the same thing as the original `jax.lax.map`, but a bit differently, so it seems analogous.

[the remark this answers is not in the document]
Comment (mattjj) [reply jax/_src/lax/control_flow/loops.py:2025]: So really concretely: we could adjust this PR to not introduce a new function, but rather add an option like `chunk_size: int | None = None` to `jax.lax.map`.

[the remark this answers is not in the document]
Comment (shoyer) [reply jax/_src/lax/control_flow/loops.py:2025]: `batch_vmap` also solves the unrelated problem that `lax.map` and `vmap` have different APIs, which makes it hard to switch between them. In practice, I think the `vmap` API is more useable.

So my preference would be to add a `batch_size` argument onto `jax.vmap` rather than a `chujnk_size `argument onto `lax.map`.

[the remark this answers is not in the document]
Comment (froystig) [reply jax/_src/lax/control_flow/loops.py:2025]: > `batch_vmap` also solves the unrelated problem [...]

If this is indeed a problem, then maybe we should consider solving it separately?

[the remark this answers is not in the document]
Comment (shoyer) [reply jax/_src/lax/control_flow/loops.py:2025]: > > `batch_vmap` also solves the unrelated problem [...]
> 
> If this is indeed a problem, then maybe we should consider solving it separately?

We could certainly add a new API for the `lax.map` functionality with `vmap` syntax, but given that it's almost equivalent to `batched_vmap` with `batch_size=1` (aside from unnecessary batching over a size one dimension), I'm not sure it's worth the trouble.

We have four options for adding support for a batched map into JAX:

1. Make `vmap`support batching (by adding a `batch_size` parameter)
2. Make `lax.map` support batching (by adding a `chunk_size` parameter)
3. Add a new function with the syntax of `vmap` (`batched_vmap`)
4. Add a new function with the syntax of `lax.map` (no current proposal)

All of these support the same essential functionality, but I think (1) and (3) are by far the most user friendly.

[the remark this answers is not in the document]
Comment (froystig) [reply jax/_src/lax/control_flow/loops.py:2025]: I'm partial to (1) or (2). I'm not sure that we need a new function, considering how close this is to two existing ones.

File: jax/_src/lax/control_flow/loops.py (hunk 2 above)
Comment (BrandonSmithJ) [reply jax/_src/lax/control_flow/loops.py]: Adding on to this, allowing a batch size that doesn't evenly divide the dimension would be very useful. I had to solve this problem in my own work, and ended up just running the divisible portion via scan and the remainder separately. While that might not be a perfect solution since it's two separate calls, it at least allows seamless usage:

```python
    def batched_vmap(f, batch_size: int, axis: int = 0):
        vmap_f = jax.vmap(f)
        scan_f = lambda carry, x: (None, vmap_f(x))

        def wrapper(array):
            # Separate into [non-]divisible parts, and reshape for scan
            array = jnp.moveaxis(array, axis, 0)
            batch_count = len(array) // batch_size
            array,extra = jnp.split(array, [batch_count * batch_size])
            array = array.reshape((batch_count, batch_size) + array.shape[1:])

            # Scan over the divisible section, and separately apply vmap to extra
            array = jax.lax.scan(scan_f, None, array)[1]
            extra = vmap_f(extra)

            # Recombine the two parts, and restore the axis position
            array = array.reshape((-1,) + array.shape[2:])
            array = jnp.append(array, extra, axis=0)
            return jnp.moveaxis(array, 0, axis)
        return wrapper
```

I'm certainly not an expert with jax, though - so I might be missing some reasons why that approach shouldn't be used.

File: tests/api_test.py [hunk 7]
```diff
@@ -10429,6 +10429,52 @@ def vmap_ref(xs, y):
...
+    ("12", 8),
+    ("16", 16),
+  )
+  def test_batch_map_pytrees(self, batch_size: int):
+    f = lambda x: {'b': x['a'] ** 2}
+    inputs = {'a': np.arange(16)}
+    expected = np.arange(16) ** 2
+
+    outputs = jax.lax.map(f, inputs, batch_size=batch_size)
+    np.testing.assert_array_equal(outputs['b'], expected)
```
Comment (mattjj) [inline tests/api_test.py:10460]: Use `self.assertAllClose(outputs['b'], expected, ...)` instead of `np.testing.assert_array_equal` (see other tests for examples).

File: jax/_src/lax/control_flow/loops.py [hunk 8]
```diff
@@ -2157,8 +2157,46 @@ def fori_loop(lower, upper, body_fun, init_val):
...
+  return scan_tree, remainder_tree
+
 @api_boundary
-def map(f, xs):
+def map(
+  f,
+  xs,
+  *,
+  in_axes: int | None | Any = 0,
+  out_axes: Any = 0,
```
Comment (mattjj) [inline jax/_src/lax/control_flow/loops.py:2197]: Can we not add in_axes/out_axes in this PR? They're complicated (e.g. it's not clear that we should support `out_axes=None`), implemented with a transpose that the calling code could do (rather than slicing along the given axes), and the test coverage here isn't good. Let's just make this PR about `batch_size`.

File: jax/_src/lax/control_flow/loops.py [hunk 9]
```diff
@@ -2157,8 +2157,46 @@ def fori_loop(lower, upper, body_fun, init_val):
...
+  else:
+    move_axes = lambda leaf, axis: jax.numpy.moveaxis(leaf, 0, axis)
+
+  leaves = [
+    move_axes(leaf, axis)
+    for leaf, axis in zip(leaves, flat_axes)
+  ]
+  return treedef.unflatten(leaves)
+
+def _batch_and_remainder(x, batch_size: int):
```
Comment (mattjj) [inline jax/_src/lax/control_flow/loops.py:2175]: Can you reuse the utilities `_partition_leading` and `_combine_leading` that are used in the scan unrolling implementation?

**Comment by shoyer** [comment]
I'm curious about the reasoning for putting this into map vs vmap? I do think the later is a more accessible/widely used API, but maybe that also makes it higher risk?

**Comment by cgarciae** [comment]
@shoyer There as pros and cons for both but not to much. `map` is a safer option.

File: jax/_src/lax/control_flow/loops.py (hunk 9 above)
Comment (cgarciae) [reply jax/_src/lax/control_flow/loops.py:2077]: These functions appear to be deleted after a synced with `main`. Keeping current implementation.

**Review by mattjj — APPROVED** [review]

**Comment by PhilipVinc** [comment]
@cgarciae FYI @inailuig noticed a while ago that the approach you are using here (which we also are using) breaks sharding propagation if you chunk an axis that is sharded.

to fix it we had to resort to shard map and sharding constraints. 

it would be nice if this could be done automatically by Jax…

**Comment by shoyer** [comment]
> @cgarciae FYI @inailuig noticed a while ago that the approach you are using here (which we also are using) breaks sharding propagation if you chunk an axis that is sharded.

+1 it would be really nice to implement `shard(map)` as `scan(shard(vmap(...)))`. This could potentially be a follow-on improvement, though...

File: jax/_src/lax/control_flow/loops.py [hunk 10]
```diff
@@ -2102,16 +2123,48 @@ def map(f, xs):
...
   """
-  g = lambda _, x: ((), f(x))
-  _, ys = scan(g, (), xs)
+  if batch_size is not None:
+    scan_xs, remainder_xs = _batch_and_remainder(xs, batch_size)
+    g = lambda _, x: ((), jax.vmap(f)(x))
+    _, scan_ys = scan(g, (), scan_xs)
+    remainder_ys = jax.vmap(f)(remainder_xs)
+    # flatten = lambda x: x.reshape(-1, *x.shape[2:])
+    flatten = lambda x: x.reshape(-1, *x.shape[2:])
```
Comment (ayaka14732) [inline jax/_src/lax/control_flow/loops.py:2161]: Duplicate?

**Review by ayaka14732 — APPROVED** [review]

File: jax/_src/lax/control_flow/loops.py (hunk 10 above)
Comment (cgarciae) [reply jax/_src/lax/control_flow/loops.py]: Removed

**Comment by gbuzzard** [comment]
I just opened an issue on this:  [jax.lax.map result changes drastically with large batch_size #27591](https://github.com/jax-ml/jax/issues/27591)

I can avoid it for now by using a smaller batch size, but it would be good to know what's going on.

**Comment by gbuzzard** [comment]
After replacing the batched map with a vmap, it appears that the bug is in vmap.
